### ISSUE TITLE
jd-gui does not run since Big Sur

### ISSUE THREAD

**Issue by rgov** [Homebrew/homebrew-cask#161798]
### Verification

- [X] I understand that [if I ignore these instructions, my issue may be closed without review](https://github.com/Homebrew/homebrew-cask/blob/master/doc/faq/closing_issues_without_review.md).
- [X] I have retried my command with `--force`.
- [X] I ran `brew update-reset && brew update` and retried my command.
- [X] I ran `brew doctor`, fixed as many issues as possible and retried my command.
- [X] I have checked the instructions for [reporting bugs](https://github.com/Homebrew/homebrew-cask#reporting-bugs).
- [X] I made doubly sure this is not a [checksum does not match / SHA256 mismatch](https://docs.brew.sh/Common-Issues#cask---checksum-does-not-match) error (do not open an issue before trying to open a PR to fix first).

### Description of issue

I installed the JD-GUI cask but the Java runtime dependency was not installed:

> ERROR launching 'JD-GUI'
> 
> No suitable Java version found on your system!
> This program requires Java 1.8+
> Make sure you install the required Java version.

The [formula](https://github.com/Homebrew/homebrew-cask/blob/629dd2adeabe3a7ac3cd8286eb5333af290f7cc6/Casks/j/jd-gui.rb) does not reference the dependency on Java.

It should probably have a `caveats` clause with `depends_on_java`, but this isn't sufficient.

I did already have the `temurin` package, version `20.0.2,9` installed, which didn't seem to work, nor did upgrading to `21.0.1,12`.

### Command that failed

brew install --cask jd-gui

### Output of command with `--verbose --debug`

```shell
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading jd-gui
==> Cask::Installer#install
==> Printing caveats
==> Cask::Installer#fetch
==> Downloading https://github.com/java-decompiler/jd-gui/releases/download/v1.6.6/jd-gui-osx-1.6.6.tar
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.1.24\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 14.1.2\)\ curl/8.1.2 --header Accept-Language:\ en --retry 3 --fail --location --silent --head https://github.com/java-decompiler/jd-gui/releases/download/v1.6.6/jd-gui-osx-1.6.6.tar
Already downloaded: /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
==> Checking quarantine support
/usr/bin/env /usr/bin/xattr -h
/usr/bin/env /usr/bin/swift -target arm64-apple-macosx14 /opt/homebrew/Library/Homebrew/cask/utils/quarantine.swift
==> Quarantine is available.
==> Verifying Gatekeeper status of /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
/usr/bin/env /usr/bin/xattr -p com.apple.quarantine /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
==> /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar is quarantined
==> Verifying checksum for 'fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar'
==> Installing Cask jd-gui
==> Cask::Installer#stage
==> Extracting primary container
==> Using container class UnpackStrategy::Tar for /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
/usr/bin/env tar --extract --no-same-owner --file /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar --directory /private/tmp/d20231205-25230-3o0o9
/usr/bin/env cp -pR /private/tmp/d20231205-25230-3o0o9/jd-gui-osx-1.6.6/. /opt/homebrew/Caskroom/jd-gui/1.6.6/jd-gui-osx-1.6.6
==> Verifying Gatekeeper status of /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
/usr/bin/env /usr/bin/xattr -p com.apple.quarantine /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
==> /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar is quarantined
==> Propagating quarantine from /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar to /opt/homebrew/Caskroom/jd-gui/1.6.6
/usr/bin/env /usr/bin/xattr -p com.apple.quarantine /Users/rzg/Library/Caches/Homebrew/downloads/fe7da8805b9556958179f399831e79761541d4d21930d561151f4d1832444a30--jd-gui-osx-1.6.6.tar
/usr/bin/env /usr/bin/xargs -0 -- /bin/chmod -h u+w
/usr/bin/env /usr/bin/xargs -0 -- /usr/bin/xattr -w com.apple.quarantine 0181\;65700b6f\;Homebrew\\x20Cask\;42FD67D0-1EA5-4F31-AEBC-4D2EAE6DDA76
==> Creating metadata directory: /opt/homebrew/Caskroom/jd-gui/.metadata/1.6.6/20231206055353.462
==> Creating metadata subdirectory: /opt/homebrew/Caskroom/jd-gui/.metadata/1.6.6/20231206055353.462/Casks
==> Installing artifacts
==> Installing artifact of class Cask::Artifact::App
==> Moving App 'JD-GUI.app' to '/Applications/JD-GUI.app'
🍺  jd-gui was successfully installed!
```

### Output of `brew doctor` and `brew config`

```shell
HOMEBREW_VERSION: 4.1.24
ORIGIN: https://github.com/Homebrew/brew
HEAD: ab6a7ed5a122b6d4f9d441ef76c86032398ced5f
Last commit: 5 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 96429d0c4d7356e72396b83523ea4c74a4eb3253
Core tap last commit: 7 months ago
Core tap branch: colmap
Core tap JSON: 06 Dec 05:48 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.Bk40vCSTFO/org.xquartz:0
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1.2-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Output of `brew tap`

```shell
homebrew/cask
homebrew/core
homebrew/services
smokris/getwindowid
yt-dlp/taps
```

**Comment by rgov** [comment]
The main executable for JD-GUI is the script at `JD-GUI.app/Contents/macOS/universalJavaApplicationStub.sh`. This script is failing to detect the Java runtime, and throwing the aforementioned error dialog.

The script reads the `JavaX/JVMVersion` key from `JD-GUI.app/Contents/Info.plist`, which contains `1.8+`. Then it invokes `/usr/libexec/java_home -F -v 1.8+` which errors that the specified version is not found.

It's unclear to me if `+` is really supported in the `java_home` utility. For instance `-v 21` works, but `-v 21+` does not.

**Comment by rgov** [comment]
It's hacky but I changed the required Java version to 21 to see if it would run with the newer JRE. I quickly hit [this exception](https://stackoverflow.com/questions/75920482/cannot-access-class-com-apple-eawt-application-in-module-java-desktop-because), but the workaround of adding `--add-exports java.desktop/com.apple.eawt=ALL-UNNAMED` to the `java` command invocation by the `universalJavaApplicationStub.sh` script worked OK.

In summary, this program is broken on modern macOS, but can be fixed with some patches to the included `universalJavaApplicationStub.sh`. Someone has shown how here:

https://github.com/g1a55er/jd-gui/commit/5e743011a932b287428dd6451e08637df2f3d9fc

While the project is very useful and ought to remain in Homebrew, unfortunately it is not actively maintained, so this workaround would have to be implemented in the formula.

**Comment by bevanjkay** [comment]
@rgov `homebrew-cask` simply distributes the packages/binaries as provided by upstream. Because we do not build anything, there is not way or precedent from patches to be applied.
The way forward is to see if the fixes can be released by the upstream developers.

**Comment by rgov** [comment]
That's a fair point, I didn't think about casks not being patched.

**Comment by rgov** [comment]
Unfortunately maybe that means the cask should be pulled. It doesn't run as-is.

**Comment by krehel** [comment]
Looks like there hasn't been an update in 4 years (last release 2019-12-25).  

If this is broken without a reasonable workaround and it affects everyone who uses it, I could support removal.

@rgov - since you know more than we do in this instance, when you say it "doesn't run as-is", can you help us understand the breadth of non-functionality?  Would it be all recent macOS versions, all Intel, ARM, etc?  We only want to pull if this is legitimately broken for an overwhelming majority of users who would download it.

**Comment by rgov** [comment]
I believe it has been broken since Big Sur, and that this brokenness affects both architectures.

**Comment by rgov** [comment]
https://github.com/java-decompiler/jd-gui/issues?q=is%3Aissue+%22Mac%22+OR+%22macOS%22+OR+%22OS+X%22+OR+%22osx%22+ Most of the recent ones refer to how `universalJavaApplicationStub.sh` is broken.

I agree it's a very useful tool and it's a shame to drop it. I think the formula could be converted to a non-cask formula that builds it with the necessary patches applied.

I installed JDK 8 using `zulu8` from homebrew/cask-versions, and then issued `JAVA_HOME=$(/usr/libexec/java_home -v 1.8) ./gradlew build` and it produced a working .app. However I don't know if Homebrew supports `depends_on` with a cask from a tap like that. Also unsure if many formulas build .app bundles like this one would.

I spent a little bit trying to get it to build with Temurin 21, but it doesn't seem to be able to run the old Gradle, and it requires too many build script changes to support newer Gradle.
